ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx

```diff
@@ -11,35 +11,167 @@
+
+  const onLegendItemClick = (seriesName: string) => {
+    setSelectedSeriesName((current) => {
+      if (current === null || current !== seriesName) {
+        return seriesName;
+      }
+      return null;
+    });
+  };
+
+  // populate series data based on query results
+  const { graphData, loading } = useMemo(() => {
+    const timeScale = getCommonTimeScale(queryResults);
+    if (timeScale === undefined) {
+      for (const query of queryResults) {
+        // does not show error message if any query is successful (due to timeScale check)
+        if (query.error) throw query.error;
+      }
+      return {
+        graphData: EMPTY_GRAPH_DATA,
+        loading: true,
+      };
+    }
+
+    const graphData: EChartsDataFormat = {
+      timeSeries: [],
+      xAxis: [],
+      legendItems: [],
+      rangeMs: timeScale.endMs - timeScale.startMs,
+    };
+    const xAxisData = [...getXValues(timeScale)];
+
+    let queriesFinished = 0;
+    for (const query of queryResults) {
+      // Skip queries that are still loading and don't have data
+      if (query.isLoading || query.data === undefined) continue;
+
+      for (const timeSeries of query.data.series) {
+        const yValues = getYValues(timeSeries, timeScale);
+        const lineSeries = getLineSeries(timeSeries.name, yValues, selectedSeriesName);
+        if (selectedSeriesName === null || selectedSeriesName === timeSeries.name) {
+          graphData.timeSeries.push(lineSeries);
+        }
+        if (legend.show && graphData.legendItems) {
+          graphData.legendItems.push({
+            id: timeSeries.name,
+            label: timeSeries.name,
+            isSelected: false,
+            color: getRandomColor(timeSeries.name),
+            onClick: () => onLegendItemClick(timeSeries.name),
+          });
+        }
+      }
+      queriesFinished++;
+    }
+    graphData.xAxis = xAxisData;
+
+    if (thresholds && thresholds.steps) {
+      const defaultThresholdColor = thresholds.default_color ?? ThresholdColors.RED;
+      thresholds.steps.forEach((step: StepOptions, index: number) => {
+        const stepPaletteColor = ThresholdColorsPalette[index] ?? defaultThresholdColor;
+        const thresholdLineColor = step.color ?? stepPaletteColor;
+        const stepOption: StepOptions = {
+          color: thresholdLineColor,
+          value: step.value,
+        };
+        const thresholdName = step.name ?? `Threshold ${index + 1} `;
+        // TODO: switch back to markLine once alternate tooltip created
+        const thresholdData = Array(xAxisData.length).fill(step.value);
+        const thresholdLineSeries = getLineSeries(thresholdName, thresholdData, selectedSeriesName, stepOption);
+        graphData.timeSeries.push(thresholdLineSeries);
+      });
+    }
+
+    return {
+      graphData,
+      loading: queriesFinished === 0,
+      allQueriesLoaded: queriesFinished === queryResults.length,
+    };
+  }, [queryResults, thresholds, selectedSeriesName, legend]);
```

Comment:
If it's snappy enough, we might not need to worry about this optimization:

But I wonder if eventually we'll want to separate: processing that we do when we initially load data, and processing that we dod when we change the selectedSeriesName (because seems like we don't need to do as much math when we just change the selected series))

Reply:
Looking at it again, I can probably pull out the thresholds logic but I'm not sure how separating out selectedSeriesName would work 🤔

The way I'm thinking about it currently, when a legend item is clicked, it does need to reprocess what gets [pushed into graphData.timeSeries](https://github.com/perses/perses/blob/a9962cbeaab6744701d0e7fefefae59d407a5aca/ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx#L88-L91) as well so that only the correct selected series line shows and for the legend and tooltip to stay in sync (that's actually what is broken about the current ECharts legend, where clicking an item does not also filter the series data that gets pushed into our custom tooltip)